### ISSUE TITLE
Missing extension in protobuf-3.5.1 + proto2

### ISSUE THREAD

**Issue by WindLeeWT** [protocolbuffers/protobuf#4293]
### We're trying to migrate from protobuf-2.6.1 to protobuf-3.5.1. 

After reading through official documents, we decide to use **protobuf-3.5.1 + proto2** which should be fully compatible with the existing protos within code base. It built and passed unit tests with few minor changes. However, when we put newly built services online, they kept showing "extensions not found" errors. After some investigation I wrote a simple example below to illustrate :

Example code was put under a temporary directory(common/rpc/temp)：

#### example_options.proto
```protobuf
package example;

import "google/protobuf/descriptor.proto";

extend google.protobuf.ServiceOptions {
  optional int32 example_version = 20000;
}
```
#### example.proto
```protobuf
package example;

import "common/rpc/temp/example_options.proto";

option cc_generic_services = true;

message ShowRequest {
  optional string name = 1;
}

message ShowResponse {
}

service ExampleService {
  option (example_version) = 100;
  rpc Show(ShowRequest) returns (ShowResponse);
}
```
#### employee.h
```c++
#include <string>

#include "google/protobuf/descriptor.h"

namespace example {

class Employee {
 public:
  explicit Employee(const std::string& name);
  void Report();

 private:
  std::string name_;
  static const google::protobuf::MethodDescriptor* method_descriptor;
};

}
```
#### employee.cc
```c++
#include "common/rpc/temp/employee.h"

#include <iostream>
#include "common/rpc/temp/example.pb.h"

namespace example {

const google::protobuf::MethodDescriptor* Employee::method_descriptor =
    nullptr;
    // Assign to nullptr: example_version found
    // Assign to "Show" method: example_version not found
    // ExampleService::descriptor()->FindMethodByName("Show");

Employee::Employee(const std::string& name) : name_(name) { }

void Employee::Report() {
  std::cout << "Employee::Report" << std::endl
            << "    Name: " << name_ << std::endl
            << "    MethodDescriptor: " << method_descriptor << std::endl;
}

}
```
#### main.cc
```c++
#include <iostream>
#include <memory>

#include "google/protobuf/descriptor.pb.h"
#include "common/rpc/temp/example.pb.h"
#include "common/rpc/temp/employee.h"

using example::example_version;
using example::ExampleService;
using example::Employee;

int main() {
  std::cout << "Enter main" << std::endl;

  const auto* descriptor = ExampleService::descriptor();
  const auto& options = descriptor->options();
  if (options.HasExtension(example_version)) {
    std::cout << "example_version found" << std::endl;
  } else {
    std::cout << "example_version not found" << std::endl;
  }

  auto e = std::make_unique<Employee>("Jack");
  e->Report();

  std::cout << "Exit main" << std::endl;
  return 0;
}
```

As mentioned in the comments, actually when **ExampleService::descriptor()** gets called directly or indirectly before main, the extension is missing. This issue seems to be related to descriptors initialization order.

I assume that extension is well supported in **protobuf-3.5.1 + proto2**, but even in proto3, custom options introduced by extensions are still supported. Please help take a look.

**Comment by xfxyjwf** [comment]
@gerben-s It seems the extension registration change in cl/162411066 (brought to opensource repo in https://github.com/google/protobuf/commit/13fd045dbb2b4dacea32be162a41d5a4b0d1802f)  breaks descriptor custom options if accessed pre-main. Can you take a look?

**Comment by orivej** [comment]
I have confirmed that this is a regression introduced in Protobuf 3.5.0.

**Comment by orivej** [comment]
@xfxyjwf Could you publish the diff of `cl/162411066` that you mentioned, to make it easier to fix this outside of Google?

**Comment by xfxyjwf** [comment]
@orivej Sorry, but I cannot publish the internal change here. You should be able to find all relevant changes in src/google/protobuf/extension_set.h, src/google/protobuf/compiler/cpp/cpp_extension.cc and src/google/protobuf/compiler/cpp/cpp_file.cc in https://github.com/google/protobuf/commit/13fd045dbb2b4dacea32be162a41d5a4b0d1802f : 
https://github.com/google/protobuf/commit/13fd045dbb2b4dacea32be162a41d5a4b0d1802f#diff-024d847982e4f522c061c45f366ab0b3
https://github.com/google/protobuf/commit/13fd045dbb2b4dacea32be162a41d5a4b0d1802f#diff-a2c26deb462bd6c9148b12e5b58b9650

I have created an internal bug and am asking the original author to take a look.

**Comment by gerben-s** [comment]
Have you tried changing the link order? Linker line should probably be written such that the order matches the dependencies.

**Comment by gerben-s** [comment]
Is this still an issue?

**Comment by orivej** [comment]
It makes sense that the issue is with the link order. I'm going to check this today.

**Comment by gerben-s** [comment]
Any updates?

**Comment by gerben-s** [comment]
@orivej Any updates?

**Comment by orivej** [comment]
Indeed, ordering compiled protoc-generated files before files that use them in global initializers solves this issue.

**Comment by georgthegreat** [comment]
TWIMC: the pre-main usage of extensions was (accidentally?) broken in protobuf=3.5, as it became dependant on the link order in case of a static linkage. It was (accidentally again) fixed in 3.15.
I believe that constinit initialization of default message instances has accidentally fixed the issue, but I can not say for sure.

I have spent a bunch of time trying to investigate this, but so far I have nothing except for the MWE with problem demonstration.
